### PR TITLE
Fix the `ClientWebSocketResponse.receive` stuck for a long time if network is suddenly unavaiable.

### LINKED ISSUE #2309: WebSocketResponse does not throw/close when connection is abruptly cut
## Long story short

This is a continuation of #961 (with some ties to #2283). Apparently, the issue is still valid (I don't know if it was ever really fixed).

I have a single websocket handler. It does two things on connection:

* Starts a sender task that sends `N` bytes over the websocket every second
* Receives (and discards) any incoming websocket messages

I connect to this handler through a simple javascript client running in Google Chrome. Everything works as expected: I receive `N` bytes every second, indefinitely.

However, when the connection is abruptly cut (by disconnecting the network cable), the websocket sender task takes 15 minutes to stop (even though the heartbeat timeout is only 5 seconds).

## Expected behaviour

I would expect the `send_bytes` operation to throw *or* the `async for msg in ws` loop to terminate/throw.

## Actual behaviour

The `send_bytes` keep buffering up writes (eventually causing the process to run out of memory and crash). Furthermore, the `async for msg in ws` loop takes 15 minutes to terminate/throw.

## Steps to reproduce

The websocket server:

	#!/usr/bin/env python3.6
	import aiohttp
	import asyncio
	import logging
	from aiohttp import web

	logger = logging.getLogger()
	logger.setLevel(logging.INFO)

	async def sender(ws):
		try:
			logging.info('sender begin')
			while True:
				# Sender never stops on disconnect for N = 10000
				N = 100000
				# Sender stops on disconnect for N = 10
				#N = 10
				logging.info(f'sending {N} bytes')
				try:
					ws.send_bytes(bytes(N))

					# Explicit drain will stall the sender on disconnect
					# but not throw (despite of the heartbeat)
					#await ws.drain()

					# The only fix is to query transport.is_closing()
					# This stops the transport as soon as the heartbeat times out
					#if ws._req.transport.is_closing():
					#	logging.info(f'sender transport is closing')
					#	break
				except Exception as error:
					logging.info(f'sender error:', exc_info=error)
				await asyncio.sleep(1.0)
		# Catch all sender Exceptions
		except Exception as error:
			logging.info('sender error:', type(error))
		finally:
			logging.info('sender end')

	async def handler(request):
		# Note the hearbeat is set to 5 seconds
		ws = web.WebSocketResponse(heartbeat=5.0)
		await ws.prepare(request)

		# Launch send task
		sender_task = asyncio.ensure_future(sender(ws))

		# Pass any message
		logging.info('receiver begin')
		try:
			async for msg in ws:
				pass
		# Catch all receiver Exceptions
		except Exception as error:
			logging.info(f'receiver error:', exc_info=error)
		finally:
			logging.info('receiver end')

		# Cancel sender task
		sender_task.cancel()

		return ws

	app = web.Application()
	app.router.add_route('GET', '/', handler)
	web.run_app(app, port=5006)

The javascript client (wrapped in HTML for browser use):

	<html>
	    <script>
	        ws = new WebSocket('ws://192.168.1.111:5006/');
	        ws.onmessage = (message) => console.log(message);
	        ws.onerror = (error) => console.error(error);
	        ws.onclose = () => console.log('closed');
	    </script>
	</html>

Note that this only happens for large `N` (around 10000). Everything works fine for say `N = 10`. I don't know why. Maybe some internal caching/buffering is causing this issue?

Note that using `await ws.drain()` only stalls the sender. This way, the process doesn't run out of memory but the connection is still in limbo.

The practical fix is to query ` ws._req.transport.is_closing()` (as pointed out by ttm56p in #2283).

I think that aiohttp should handle this internally by whatever means necessary.

It took a long time for me to track down a memory leak to this issue. The writes just kept buffering up. Note that this also happens if you simply let a browser tab sit idle for a long time (apparently, Chrome will stop sending websocket data after a period of inactivity in the tab). In practice, I saw my websocket servers crashing every now and then.

## Your environment

Aiohttp 2.2.3 and Python 3.6.1. Google Chrome 61.0.

### PR DESCRIPTION
**Pull request by Hanaasagi**
## What do these changes do?

This PR fix the websocket client will get stuck when a the internect is unavaiable.

## How to reproduce it

Run following code, and turn off the internet after seeing the `Connected` text in the console.

```Python
from asyncio import run
from aiohttp import ClientSession

async def main():
    async with ClientSession() as session:
        async with session.ws_connect('https://echo.websocket.org/',
                                      heartbeat=10) as ws:
            print('Connected', ws)
            print('Received', await ws.receive())

run(main())
```

In my local machine, it will block ablout 8 min and get a `CLOSED` message.

## Explain Why 

`receive` method has a `timeout` argument, and the default value is `None`. It means no timeout for reading from a connection. The websocket pong frame and user data frame uses the same timeout. So if `aiohttp` send a ping frame successfully, then we turn off the network. The `receive` coroutine will be block at the `msg = await self._reader.read()`. Because there are no data comes from the server.

https://github.com/aio-libs/aiohttp/blob/184274d9b28bbfa06ac60e48bf286a761c6a6cb0/aiohttp/client_ws.py#L213-L238

Even if we call `_pong_not_received`, it doesn't cancel the `receive` coroutine.

https://github.com/aio-libs/aiohttp/blob/184274d9b28bbfa06ac60e48bf286a761c6a6cb0/aiohttp/client_ws.py#L101-L106

The `receive` coroutine finished until a `EofStream` error from the TCP Stack.

## What did this PR do 

Add a simple flag(`asyncio.Event`) `_is_heartbeat_failed` in `ClientWebSocketResponse`. And when call `ClientWebSocketResponse.receive`, it both await `self._reader.read()` and this flag by `asyntio.wait` . So if the heartbeat is failed, the `ClientWebSocketResponse.receive` will be awake.

## Related issue number

Fix https://github.com/aio-libs/aiohttp/issues/2309

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by webknjaz** [comment]
Please fill out the PR template with the details on what this is supposed to do.

File: aiohttp/client_ws.py [hunk 1]
```diff
@@ -235,11 +239,32 @@ async def receive(self, timeout: Optional[float] = None) -> WSMessage:
...
+                        )
+                        # If server doesn't pong, but return data normally,
+                        # supress the exception.
+                        if read_task in done:
+                            is_heartbeat_failed_task.cancel()
+                            msg = read_task.result()
+                        elif is_heartbeat_failed_task in done:
+                            read_task.cancel()
+                            assert self._exception is not None
+                            raise self._exception
```
Comment (Hanaasagi) [inline aiohttp/client_ws.py:266]: I'm not sure that raise ` TimeoutError` or just return a `WSMessage(WSMsgType.ERROR, ..)` message which is better.

**Comment by gil-cohen** [comment]
any update?

File: aiohttp/client_ws.py [hunk 2]
```diff
@@ -235,11 +239,32 @@ async def receive(self, timeout: Optional[float] = None) -> WSMessage:
 
             try:
                 self._waiting = self._loop.create_future()
+
+                read_task: asyncio.Task[WSMessage] = asyncio.create_task(
+                    self._reader.read()
+                )
+                is_heartbeat_failed_task: asyncio.Task[bool] = asyncio.create_task(
+                    self._is_heartbeat_failed.wait()
+                )
```
Comment (astrand) [inline aiohttp/client_ws.py:248]: Do we really need two additional tasks in order to solve this bug? Is it at all necessary to check heartbeat status, or is it sufficient to check transport.is_closing()?

**Comment by Dreamsorcerer** [comment]
The linked issue was closed, would be good if @bdraco or @Hanaasagi could confirm if this PR is no longer relevant.

**Comment by bdraco** [comment]
The receive resolves immediately when we know the connection is dropped now

**Comment by bdraco** [comment]
Very likely this issue is fixed. Please let us know if you still experience this behavior on 3.10.5 or later and we can reopen.

**Pull request closed without merging**